riklet/src/core.rs

```diff
@@ -1,490 +1,175 @@
+pub enum RikletError {
+    #[error("Failed to parse workload definition: {0}")]
+    WorkloadParseError(serde_json::Error),
+
+    #[error("Message status error: {0}")]
+    MessageStatusError(tonic::Status),
+
+    #[error("Configuration error: {0}")]
+    ConfigurationError(ConfigurationError),
+
+    #[error("Failed to connect client: {0}")]
+    ConnectionError(tonic::transport::Error),
 
-// const TAP_SCRIPT_DEFAULT_LOCATION: &str = "/app/setup-host-tap.sh";
-const MASK_LONG: &str = "255.255.255.252";
-const DEFAULT_AGENT_PORT: u16 = 8080;
+    #[error("Runtime error: {0}")]
+    RuntimeManagerError(RuntimeError),
+}
+type Result<T> = std::result::Result<T, RikletError>;
 
-#[derive(Debug)]
 pub struct Riklet {
+    config: Configuration,
     hostname: String,
     client: WorkerClient<Channel>,
     stream: Streaming<InstanceScheduling>,
-    image_manager: ImageManager,
-    container_runtime: Runc,
-    workloads: HashMap<String, Vec<Container>>,
-    ip_allocator: IpAllocator,
-    function_config: FnConfiguration,
+    // Can be pod or function runtimes
+    // The key is the instance id
+    runtimes: HashMap<String, Box<dyn Runtime>>,
 }
 
 impl Riklet {
-    /// Display a banner
-    fn banner() {
-        println!(
-            r#"
-        ______ _____ _   __ _      _____ _____
-        | ___ \_   _| | / /| |    |  ___|_   _|
-        | |_/ / | | | |/ / | |    | |__   | |
-        |    /  | | |    \ | |    |  __|  | |
-        | |\ \ _| |_| |\  \| |____| |___  | |
-        \_| \_|\___/\_| \_/\_____/\____/  \_/
-        "#
-        );
-    }
-
-    /// Bootstrap a Riklet in order to run properly.
-    pub async fn bootstrap() -> Result<Self, Box<dyn Error>> {
-        event!(Level::DEBUG, "Riklet bootstraping process started.");
-        // Get the hostname of the host to register
-        let hostname = gethostname::gethostname().into_string().unwrap();
-
-        // Display the banner, just for fun :D
-        Riklet::banner();
-
-        // Load the configuration
-        let config = Configuration::load()?;
-
-        // load the function runtime configuration
-        let function_config = FnConfiguration::load();
-
-        // Connect to the master node scheduler
-        let mut client = WorkerClient::connect(config.master_ip.clone()).await?;
-        event!(Level::DEBUG, "gRPC WorkerClient connected.");
-
-        event!(Level::DEBUG, "Node's registration to the master");
-        let request = Request::new(WorkerRegistration {
-            hostname: hostname.clone(),
-        });
-        let stream = client.register(request).await?.into_inner();
-
-        event!(Level::DEBUG, "Container runtime initialization");
-        let container_runtime = Runc::new(config.runner.clone())?;
-        event!(Level::DEBUG, "Image manager initialization");
-        let image_manager = ImageManager::new(config.manager.clone())?;
+    async fn handle_workload(&mut self, workload: &InstanceScheduling) -> Result<()> {
```

Comment:
In case this method fails by any way, at any point, it will cause an error elevation which will terminate riklet. It would make sense to have a better way to handle this, so that we have a resilient way to handle workloads.